arch/arc/core/cpu_idle.S

```diff
@@ -43,7 +43,18 @@ SECTION_FUNC(TEXT, z_arch_cpu_idle)
 
 	ld r1, [z_arc_cpu_sleep_mode]
 	or r1, r1, (1 << 4) /* set IRQ-enabled bit */
+/*
+ * It's found that (in nsim_hs_smp and hsdk), when cpu
```

Comment:
yikes. so this means that idle threads are just going to spin wasting power? is there a bug filed on this?

Reply:
Let me check what's wrong here. I suggest to report it as a separate issue and continue discussion there so we track it for the next release etc.

Reply:
Interrupts should wake up the processor. Didn't check in detail, but is the z_arc_cpu_sleep_mode initialized correctly? It may set the threshold to mask the IPI interrupt.

Reply:
@ruuddw, @vonhust  `z_arc_cpu_sleep_mode` initialized with `_ARC_V2_WAKE_IRQ_LEVEL` = `_ARC_V2_DEF_IRQ_LEVEL`, see https://github.com/foss-for-synopsys-dwc-arc-processors/zephyr/blob/topic-fixes-smp/arch/arc/include/v2/irq.h.

And `_ARC_V2_DEF_IRQ_LEVEL ` in absence of `CONFIG_ARC_SECURE_FIRMWARE` = `CONFIG_NUM_IRQ_PRIO_LEVELS - 1`, i. e. =1 for `nsim_hs_smp` & `hsdk` due to `NUM_IRQ_PRIO_LEVELS` = 2, see https://github.com/foss-for-synopsys-dwc-arc-processors/zephyr/blob/topic-fixes-smp/soc/arc/snps_nsim/Kconfig.defconfig.hs_smp#L12 and https://github.com/foss-for-synopsys-dwc-arc-processors/zephyr/blob/topic-fixes-smp/soc/arc/snps_arc_hsdk/Kconfig.defconfig#L19.

Inter-Core Interrupts are initialized with `ARCV2_ICI_IRQ_PRIORITY = 1`, see https://github.com/foss-for-synopsys-dwc-arc-processors/zephyr/blob/topic-fixes-smp/arch/arc/core/arc_smp.c#L24.

Now PRM says:
> Any interrupt with a higher priority than this interrupt threshold awakens the processor.

But later in an example I see:
>  SLEEP 0x014: Enter sleep and enable interrupts and sets interrupt priority level to 4. Any interrupt with a priority EQUAL TO or HIGHER THAN 4 awakens the processor.

Also I know that in the Linux kernel we rely on EQUAL TO part, i.e. we specify the IRQ level as when initializing all IRQs, see https://elixir.bootlin.com/linux/latest/source/arch/arc/kernel/process.c#L109 and that works. So I believe there should be something else.

Reply:
Thanks for checking - if the sleep works for Linux, it should work for Zephyr as well (i.e. not HW issue).

Reply:
FWIW: we merged SMP with a generic spin-for-idle implementation, and it was left that way for several months until the IPI framework landed.  And it was that change that broke SMP on ARC and is being addressed here.  Technically this isn't even a regression.

We should track it with a bug and fix it, but I don't see why this implementation can't merge.

Reply:
For HW, e.g., HSDK,  sleep now works as expected  after i fixed the bug in systick driver (ipi interrupt is always preempted by sys tick int).  But in nsim_hs_smp, sleep still doesn't work as expected, it may be a nsim issue. Considering nsim is an instruction simulator,  there are few impact of using loop to replace sleep.